### DISCUSSION TITLE
Suspend vs Freeze

### DISCUSSION THREAD

**Discussion by anonymous1184** [winsiderss/systeminformer#1854] [Questions / Answers] [answered]
What's the difference between `Suspend` and `Freeze`?

Suspend I see is from `NtSuspendProcess()` but freeze is `PhFreezeProcess()`.

**Comment by dmex** [comment]
> Suspend I see is from NtSuspendProcess() but freeze is PhFreezeProcess().

What does PhFreezeProcess call instead of NtSuspendProcess?

**Reply by anonymous1184** [reply]
> What does PhFreezeProcess call instead of NtSuspendProcess?

Sure, but also, what's the difference between the two states?

For example, a suspended process is a process whose threads are in wait status and the wait reason is *suspended*.

What is a frozen process?

**Reply by diversenok** [reply, marked as answer]
Freezing is a stronger version of suspension that cannot be undone via resuming (only thawing). Effectively, when a process is frozen, it adds an additional (permanent) +1 to the suspension counter of each thread. The current implementation of `PhFreezeProcess` uses state change objects, so it performs freezing, not suspension.

If you're interested in more details, I have a [repository about this topic](https://github.com/diversenok/Suspending-Techniques). It is slightly outdated (doesn't cover freeze-bypassing threads, for instance) but can still shed some light on the topic.

**Reply by anonymous1184** [reply]
Thanks a lot, the answer is perfectly clear and that repo is exactly the source of information anyone should ever need.
